codec/src/test/java/io/netty/handler/codec/DelimitedLengthFieldBasedFrameDecoderTest.java

```diff
@@ -0,0 +1,185 @@
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+public class DelimitedLengthFieldBasedFrameDecoderTest {
+
+    private static List<List<Byte>> getRandomByteSlices(byte[] bytes) {
+        List<Byte> byteList = Bytes.asList(bytes);
+
+        int numSlices = nextInt(2, 10);
+        List<Integer> sliceIndexes = new ArrayList<Integer>();
+        for (int i = 0; i < numSlices; i++) {
+            sliceIndexes.add(nextInt(0, bytes.length));
+        }
+        Collections.sort(sliceIndexes);
+
+        List<List<Byte>> slices = new ArrayList<List<Byte>>(numSlices);
+
+        int byteInd = 0;
+        for (int sliceIndex : sliceIndexes) {
+            // System.out.println(String.format("Slice from %d through %d", byteInd, sliceIndex));
+            slices.add(byteList.subList(byteInd, sliceIndex));
+            byteInd = sliceIndex;
+        }
+        // System.out.println(String.format("Slice from %d through %d", byteInd, byteList.size()));
+        slices.add(byteList.subList(byteInd, byteList.size()));
+
+        return slices;
+    }
+
+    /**
+     * Generates a new pseudo-random integer within the specific range.
+     *
+     * This is essentially the same method that is present in Apache commons-lang.  It is simply copied here to avoid
+     * bringing in a new dependency
+     *
+     * @param startInclusive the lowest value that can be generated
+     * @param endExclusive
+     *
+     * @return a pseurandom number in [startInclusive, endExclusive)
+     */
+    private static int nextInt(final int startInclusive, final int endExclusive) {
+        if (startInclusive == endExclusive) {
+            return startInclusive;
+        }
+
+        return startInclusive + ThreadLocalRandom.current().nextInt(endExclusive - startInclusive);
+    }
+
+    @Test
+    public void delimitedLengths() throws Exception {
+        Charset charset = CharsetUtil.ISO_8859_1;
+        EmbeddedChannel ch = getTestChannel(charset, 100, 0, false);
+
+        String v1 = "a";
+        String v2 = "abcdefghij";
+        String v3 =
+                "abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuv" +
+                "wxyz";
+
+        writeStringAndAssert(ch, v1, charset, false, false);
+        writeStringAndAssert(ch, v2, charset, false, false);
+        writeStringAndAssert(ch, v3, charset, false, true);
+
+        writeStringAndAssert(ch, v1, charset, true, false);
+        writeStringAndAssert(ch, v2, charset, true, false);
+        writeStringAndAssert(ch, v3, charset, true, true);
+
+        writeStringAndAssert(ch, v1, charset, false, false);
+
+        Assert.assertFalse(ch.finishAndReleaseAll());
+    }
+
+    @Test
+    public void multipleFramesSingleBuffer() throws Exception {
+        Charset charset = CharsetUtil.ISO_8859_1;
```

Comment:
It makes sense also to testing with non one-byte encoding (for example UTF-16/UTF-8).